third_party/amd/lib/TritonAMDGPUToLLVM/TargetInfo.cpp

```diff
@@ -103,34 +105,183 @@ Value TargetInfo::loadDShared(RewriterBase &rewriter, Location loc, Value ptr,
+
+  if (!valType.isIntOrIndex()) {
+    val = bitcast(val, int_ty(originalBits));
+    actualType = int_ty(originalBits);
+  }
+
+  if (originalBits < bits) {
+    val = sext(int_ty(bits), val);
+    actualType = int_ty(bits);
+  }
+
+  return actualType;
+}
+
+static inline void castFromInt(RewriterBase &rewriter, Location loc, Value &val,
+                               Type valType, unsigned bits) {
+  unsigned originalBits = valType.getIntOrFloatBitWidth();
+
+  if (originalBits < bits) {
+    val = trunc(int_ty(originalBits), val);
+  }
+
+  if (!valType.isIntOrIndex()) {
+    val = bitcast(val, valType);
+  }
+}
+
 bool TargetInfo::warpReduce(RewriterBase &rewriter, Location loc,
                             SmallVector<Value> &acc, triton::ReduceOp op,
                             unsigned numLaneToReduce,
                             unsigned interleave) const {
-  return false;
+  if (numLaneToReduce != 64)
+    return false;
+
+  if (auto family = getISAFamily();
+      family != ISAFamily::CDNA3 && family != ISAFamily::CDNA2) {
+    return false;
+  }
+
+  Operation *reduxOp = getSingleCombinerFromReduceOp(op);
+  if (!reduxOp)
+    return false;
+
+  auto createDppReduxOp = [&](Type valType, Value &src, int dppCtrl,
+                              int rowMask, int bankMask,
+                              bool boundCtrl) -> Value {
+    // DPP has limited support for data types, so here we need to
+    // cast non-integer types or integer types shorter than 32 bits
+    // to int32, except for fp32.
+    Type actualType = valType;
+    if (!valType.isF32()) {
+      actualType = castToInt(rewriter, loc, src, valType, 32);
+    }
+
+    Value dppResult =
+        rewriter
+            .create<ROCDL::DPPUpdateOp>(loc, actualType, src, src,
+                                        rewriter.getI32IntegerAttr(dppCtrl),
+                                        rewriter.getI32IntegerAttr(rowMask),
+                                        rewriter.getI32IntegerAttr(bankMask),
+                                        rewriter.getBoolAttr(boundCtrl))
+            .getRes();
+
+    if (!valType.isF32()) {
+      castFromInt(rewriter, loc, src, valType, 32);
+      castFromInt(rewriter, loc, dppResult, valType, 32);
+    }
+
+    IRMapping mapping;
+    mapping.map(reduxOp->getOperand(0), src);
+    mapping.map(reduxOp->getOperand(1), dppResult);
+    return rewriter.clone(*reduxOp, mapping)->getResult(0);
+  };
+
+  for (int i = 0; i < acc.size(); i++) {
+    Value buf;
+    auto valType = acc[i].getType();
+
+    /*
+      Here's the implementation of full-wavefront reduction using dpp.
+      https://gpuopen.com/learn/amd-gcn-assembly-cross-lane-operations/
+
+      Each step has a v_mov_dpp instruction following the redux op. In
+      some cases, the lower-level compiler could merge them into single
+      instruction. For example, v_mov_dpp + max => v_max_dpp.
+
+      In DPP, each row consists of 16 consecutive lanes.
```

Comment:
I'd suggest something like

"For gfx9, we have 64 threads per warp. These 64 threads are arranged into 4 rows, with each row being 16 threads. Each 16 threads are arranged further into 4 banks, with each bank being 4 threads. Overall it's in a (row, bank, thread) structure. When shuffling, we use row/bank mask to indicate which row/bank to participate. Then modifier like `row_shr` and `row_bcast` means exact data movement schemes. In the following instructions, taking row 0 as an example:"